src/job_cache/job_cache.cpp

```diff
@@ -215,49 +239,95 @@ FindJobResponse Cache::read(const FindJobRequest &find_request) {
   MessageParser parser(socket_fd.get());
   std::vector<std::string> messages;
 
-  while (true) {
-    MessageParserState state = parser.read_messages(messages);
+  MessageParserState state = MessageParserState::Continue;
+  do {
+    state = parser.read_messages(messages);
 
     if (state == MessageParserState::StopFail) {
-      // TODO: Try to reconnect to the daemon, launching our own if need be.
-      //       if that fails then depending on user preference either fail
-      //       or return a cache miss.
-      wcl::log::fatal("Cache::read(): failed receiving message");
+      wcl::log::error("Cache::read(): failed receiving message");
+      return wcl::result_error<FindJobResponse>(FindJobError::FailedMessageReceive);
     }
 
     if (state == MessageParserState::StopSuccess && messages.empty()) {
-      // TODO: Try to reconnect to the daemon, launching our own if need be.
-      //       if that fails then depending on user preference either fail
-      //       or return a cache miss.
-      wcl::log::fatal("Cache::read(): daemon exited without responding");
-    }
-
-    // MessageParser tries to avoid this but we should defend against
-    // the case where no error has yet occured but messages is still empty.
-    if (state == MessageParserState::Continue && messages.empty()) {
-      continue;
+      wcl::log::error("Cache::read(): daemon exited without responding");
+      return wcl::result_error<FindJobResponse>(FindJobError::NoResponse);
     }
 
-    if (messages.size() != 1) {
+    if (messages.size() > 1) {
       wcl::log::info("message.size() == %lu", messages.size());
       for (const auto &message : messages) {
         wcl::log::info("message.size() = %lu, message = '%s'", message.size(), message.c_str());
       }
-      wcl::log::fatal("Cache::read(): daemon responded with too many results");
+      wcl::log::error("Cache::read(): daemon responded with too many results");
+      return wcl::result_error<FindJobResponse>(FindJobError::TooManyResponses);
     }
 
-    break;
-  }
+    // We have a singular valid response
+    if (messages.size() == 1) {
+      break;
+    }
+
+    // MessageParser tries to avoid this but we should defend against
+    // the case where no error has yet occured but messages is still empty.
+  } while (state == MessageParserState::Continue);
 
   wcl::log::info("Cache::read(): message rx: %s", messages[0].c_str());
 
   JAST json;
   std::stringstream parseErrors;
   if (!JAST::parse(messages[0], parseErrors, json)) {
-    wcl::log::fatal("Cache::read(): failed to parse daemon response");
+    wcl::log::error("Cache::read(): failed to parse daemon response");
+    return wcl::result_error<FindJobResponse>(FindJobError::FailedParseResponse);
+  }
+
+  return wcl::result_value<FindJobError>(FindJobResponse(json));
+}
+
+// TODO: track total number of times we've failed in order to
+// fail fast after some threshold when the users requests a
+// cache miss on fail
```

Comment:
We should probably do this now. Should just be a simple static counter inside Cache::read. I think 100 attempts or something like that is a fair number?